crates/primitives/snos-output/src/codec.rs

```diff
@@ -0,0 +1,136 @@
+/// EVM developer should be aware of that and prevent data loss by not using the higest 4 bits
+pub trait SnosCodec: Sized {
+    /// Returns number of field elements required to encode current snos output field
+    fn size_in_felts(&self) -> usize;
+    /// Encodes current snos output field as felt array and appends to the result
+    fn encode_to(self, output: &mut Vec<StarkFelt>);
+    /// Tries to decode snos output field given a felt reader instance
+    fn decode(input: &mut FeltReader) -> Result<Self, FeltReaderError>;
+    /// Converts structured snos program output into array of field elements
+    fn into_encoded_vec(self) -> Vec<StarkFelt> {
+        let mut output: Vec<StarkFelt> = Vec::with_capacity(self.size_in_felts());
+        self.encode_to(&mut output);
+        output
+    }
+}
+
+impl SnosCodec for StarkFelt {
+    fn size_in_felts(&self) -> usize {
+        1
+    }
+
+    fn encode_to(self, output: &mut Vec<StarkFelt>) {
+        output.push(self);
+    }
+
+    fn decode(input: &mut FeltReader) -> Result<Self, FeltReaderError> {
+        input.read()
+    }
+}
+
+impl<T: SnosCodec> SnosCodec for Vec<T> {
+    fn size_in_felts(&self) -> usize {
+        self.iter().map(|elt| elt.size_in_felts()).sum()
+    }
+
+    fn encode_to(self, output: &mut Vec<StarkFelt>) {
+        let segment_size = self.size_in_felts() as u64;
+        output.push(segment_size.into());
+        for elt in self.into_iter() {
+            elt.encode_to(output);
+        }
+    }
+
+    fn decode(input: &mut FeltReader) -> Result<Self, FeltReaderError> {
+        let mut segment = input.read_segment()?;
+        let mut elements: Vec<T> = Vec::new();
+        while segment.remaining_len() > 0 {
+            elements.push(T::decode(&mut segment)?);
+        }
+        Ok(elements)
+    }
+}
+
+impl SnosCodec for MessageL2ToL1 {
+    fn size_in_felts(&self) -> usize {
+        3 + self.payload.len()
+    }
+
+    fn encode_to(self, output: &mut Vec<StarkFelt>) {
+        output.push(self.from_address);
+        output.push(self.to_address);
+        self.payload.encode_to(output);
+    }
+
+    fn decode(input: &mut FeltReader) -> Result<Self, FeltReaderError> {
+        Ok(Self { from_address: input.read()?, to_address: input.read()?, payload: Vec::<StarkFelt>::decode(input)? })
+    }
+}
+
+impl SnosCodec for MessageL1ToL2 {
+    fn size_in_felts(&self) -> usize {
+        5 + self.payload.len()
+    }
+
+    fn encode_to(self, output: &mut Vec<StarkFelt>) {
+        output.push(self.from_address);
+        output.push(self.to_address);
+        output.push(self.nonce);
+        output.push(self.selector);
+        self.payload.encode_to(output);
+    }
+
+    fn decode(input: &mut FeltReader) -> Result<Self, FeltReaderError> {
+        Ok(Self {
+            from_address: input.read()?,
+            to_address: input.read()?,
+            nonce: input.read()?,
+            selector: input.read()?,
```

Comment:
You should be calling `.decode()` for each one of those lines.
Here you are manually doing the inline of the impl of decode for StarkFelt.
But you are not supposed to know about the internal impl